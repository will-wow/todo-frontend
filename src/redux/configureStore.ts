import { applyMiddleware, createStore, Store as ReduxStore } from "redux";
import thunkMiddleware from "redux-thunk";

import rootReducer, { Store } from "./reducers";

export default function configureStore(): ReduxStore<Store> {
  const preloadedState = {};

  const store = createStore(
    rootReducer,
    preloadedState,
    applyMiddleware(thunkMiddleware)
  );

  return store;
}
