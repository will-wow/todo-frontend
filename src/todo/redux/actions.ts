import { isError } from "result-async";

import * as Todo from "../todo";
import * as Api from "../../api";

interface LoadTodoListSuccess {
  type: "loadTodoListSuccess";
  payload: {
    todoList: Todo.List;
  };
}

interface UpdateTodo {
  type: "updateTodo";
  payload: {
    todo: Todo.T;
  };
}

interface CreateTodoSuccess {
  type: "createTodoSuccess";
  payload: {
    todo: Todo.T;
  };
}

interface RemoveTodo {
  type: "removeTodo";
  payload: {
    todo: Todo.T;
  };
}

interface NewTodo {
  type: "newTodo";
}

export type TodoAction =
  | LoadTodoListSuccess
  | UpdateTodo
  | CreateTodoSuccess
  | RemoveTodo
  | NewTodo;

export const loadTodoList = () => async (
  dispatch: React.Dispatch<TodoAction>
) => {
  const response = await Api.get<Todo.List, string>("/");

  if (isError(response)) return null;

  const todoList = response.ok;

  dispatch({ type: "loadTodoListSuccess", payload: { todoList } });
  dispatch({ type: "newTodo" });
};

export const updateTodo = (todo: Todo.T) => async (
  dispatch: React.Dispatch<TodoAction>
) => {
  dispatch({ type: "updateTodo", payload: { todo: todo } });

  if (!Todo.isNew(todo)) {
    return Api.put<Todo.T, string>(`/${todo.id}`, Todo.toApi(todo));
  } else {
    const newTodo = await Api.post<Todo.T, string>("/", Todo.toApi(todo));

    if (isError(newTodo)) return null;

    const updatedTodo = { ...todo, ...newTodo.ok };

    dispatch({ type: "createTodoSuccess", payload: { todo: updatedTodo } });
    dispatch({ type: "newTodo" });
  }
};

export const deleteTodo = (todo: Todo.T) => async (
  dispatch: React.Dispatch<TodoAction>
) => {
  await Api.remove<Todo.T, string>(`/${todo.id}`);

  dispatch({ type: "removeTodo", payload: { todo } });
};
