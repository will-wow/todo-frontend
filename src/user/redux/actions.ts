interface UpdateUsername {
  type: "user::UpdateUsername";
  payload: {
    username: string;
  };
}

export type UserAction = UpdateUsername;

export const updateUsername = (username: string): UpdateUsername => ({
  type: "user::UpdateUsername",
  payload: {
    username
  }
});
