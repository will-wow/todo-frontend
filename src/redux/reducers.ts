import todo, { TodoStore } from "../todo/redux/reducer";

import { combineReducers } from "redux";

export interface Store {
  todo: TodoStore;
}

const reducers = combineReducers({ todo });

export default reducers;
