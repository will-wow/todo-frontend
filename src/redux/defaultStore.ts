import { Store } from "./reducers";

const defaultStore: Store = {
  todo: {
    todoMap: {}
  }
};

export default defaultStore;
