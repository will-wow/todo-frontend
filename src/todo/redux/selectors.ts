import * as Todo from "../todo";
import { Store } from "../../redux/reducers";

export const getTodoList = (state: Store): Todo.List =>
  Todo.mapToList(state.todo.todoMap);
