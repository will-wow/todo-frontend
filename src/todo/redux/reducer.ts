import * as Todo from "../todo";
import { TodoAction } from "./actions";
import { Reducer } from "redux";
import defaultStore from "../../redux/defaultStore";

export interface TodoStore {
  todoMap: Todo.Map;
}

const reducer: Reducer<TodoStore, TodoAction> = (
  state = defaultStore.todo,
  action
) => {
  switch (action.type) {
    case "newTodo": {
      const todo = Todo.build();
      return {
        ...state,
        todoMap: {
          ...state.todoMap,
          [Todo.key(todo)]: todo
        }
      };
    }
    case "loadTodoListSuccess": {
      return {
        ...state,
        todoMap: Todo.listToMap(action.payload.todoList)
      };
    }
    case "updateTodo": {
      const { todo } = action.payload;
      const { todoMap } = state;

      const updatedMap = { ...todoMap, [Todo.key(todo)]: todo };

      return { ...state, todoMap: updatedMap };
    }
    case "createTodoSuccess": {
      const { todo } = action.payload;
      const { todoMap } = state;

      const updatedMap = {
        ...todoMap,
        [Todo.key(todo)]: todo
      };

      return { ...state, todoMap: updatedMap };
    }
    case "removeTodo": {
      const { todo } = action.payload;
      const { todoMap } = state;

      const updatedMap = { ...todoMap };
      delete updatedMap[Todo.key(todo)];

      return { ...state, todoMap: updatedMap };
    }
    default: {
      return state;
    }
  }
};

export default reducer;
