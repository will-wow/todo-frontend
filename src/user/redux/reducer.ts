import { Reducer } from "redux";
import defaultStore from "../../redux/defaultStore";
import { UserAction } from "./actions";

export interface UserStore {
  username: string;
}

const reducer: Reducer<UserStore, UserAction> = (
  state = defaultStore.user,
  action
) => {
  switch (action.type) {
    case "user::UpdateUsername": {
      return { username: action.payload.username };
    }
    default: {
      return state;
    }
  }
};

export default reducer;
