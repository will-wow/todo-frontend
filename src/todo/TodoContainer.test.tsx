import React from "react";
import { shallow, ShallowWrapper, mount, ReactWrapper } from "enzyme";

import { Provider } from "react-redux";
import configureMockStore from "redux-mock-store";
import thunk from "redux-thunk";

import Container, { TodoContainer, TodoContainerProps } from "./TodoContainer";
import { Store } from "../redux/reducers";
import defaultStore from "../redux/defaultStore";

const middleware = [thunk];
const mockStore = configureMockStore<Store>(middleware);

const TODO_STORE = {
  todoMap: {
    1: {
      id: 1,
      title: "Pending Item",
      done: false,
      createdAt: "2019-01-01"
    },
    abc: {
      id: "",
      uuid: "abc",
      title: "New Item",
      done: false,
      createdAt: "2019-02-01"
    },
    3: {
      id: 1,
      title: "Done Item",
      done: true,
      createdAt: "2019-01-01"
    }
  }
};

describe("TodoContainer", () => {
  describe("connected component", () => {
    let wrapper: ReactWrapper;
    beforeEach(() => {
      const store = mockStore({
        ...defaultStore,
        todo: TODO_STORE
      });

      wrapper = mount(
        <Provider store={store}>
          <Container />
        </Provider>
      );
    });

    it("snapshots", () => {
      expect(wrapper.debug()).toMatchSnapshot();
    });
  });

  describe("component", () => {
    let wrapper: ShallowWrapper;
    let props: TodoContainerProps;

    beforeEach(() => {
      jest
        .spyOn(React, "useEffect")
        .mockImplementation((f: React.EffectCallback) => f());

      props = {
        todoList: [],
        loadTodoList: jest.fn(),
        updateTodo: jest.fn(),
        deleteTodo: jest.fn()
      };

      wrapper = shallow(<TodoContainer {...props} />);
    });

    it("loads the list on start", () => {
      expect(props.loadTodoList).toHaveBeenCalled();
    });

    describe("given a list", () => {
      beforeEach(() => {
        props.todoList = [
          {
            id: 1,
            title: "Test",
            done: false,
            createdAt: "2019-01-01"
          }
        ];

        wrapper = shallow(<TodoContainer {...props} />);
      });

      it("renders a list", () => {
        expect(wrapper.find("TodoList").prop("todoList")).toEqual(
          props.todoList
        );
      });
    });
  });
});
