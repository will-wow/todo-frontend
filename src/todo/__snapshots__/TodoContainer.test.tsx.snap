// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TodoContainer connected component snapshots 1`] = `
"<Provider store={{...}}>
  <Connect(TodoContainer)>
    <TodoContainer todoList={{...}} loadTodoList={[Function]} updateTodo={[Function]} deleteTodo={[Function]}>
      <div className=\\"Todo\\">
        <TodoList todoList={{...}} onTodoChange={[Function]} onDelete={[Function]}>
          <styled.div className=\\"TodoList\\" mx=\\"auto\\" style={{...}}>
            <StyledComponent className=\\"TodoList\\" mx=\\"auto\\" style={{...}} forwardedComponent={{...}} forwardedRef={{...}}>
              <div className=\\"TodoList sc-bdVaJa hTTfNc\\" style={{...}}>
                <Styled(styled.div) mb={2} justifyContent=\\"space-between\\" alignItems=\\"flex-end\\">
                  <StyledComponent mb={2} justifyContent=\\"space-between\\" alignItems=\\"flex-end\\" forwardedComponent={{...}} forwardedRef={{...}}>
                    <div className=\\"sc-bdVaJa sc-bwzfXH dpqGKp\\">
                      <Styled(Styled(styled.div)) as=\\"h2\\" m={0} fontSize={4} fontWeight=\\"bold\\">
                        <StyledComponent as=\\"h2\\" m={0} fontSize={4} fontWeight=\\"bold\\" forwardedComponent={{...}} forwardedRef={{...}}>
                          <h2 fontSize={4} fontWeight=\\"bold\\" className=\\"sc-bdVaJa sc-htpNat sc-bxivhb kXNIHy\\">
                            Todo List for:
                          </h2>
                        </StyledComponent>
                      </Styled(Styled(styled.div))>
                      <Connect(UsernameInput)>
                        <UsernameInput username=\\"User\\" updateUsername={[Function]}>
                          <Styled(Styled(styled.div)) value=\\"User\\" onChange={[Function: handleChange]} placeholder=\\"Enter your name here\\" as=\\"input\\" p={2} borderWidth=\\"1px\\" borderStyle=\\"solid\\" borderColor=\\"gray\\" fontSize={1}>
                            <StyledComponent value=\\"User\\" onChange={[Function: handleChange]} placeholder=\\"Enter your name here\\" as=\\"input\\" p={2} borderWidth=\\"1px\\" borderStyle=\\"solid\\" borderColor=\\"gray\\" fontSize={1} forwardedComponent={{...}} forwardedRef={{...}}>
                              <input value=\\"User\\" onChange={[Function: handleChange]} placeholder=\\"Enter your name here\\" fontSize={1} className=\\"sc-bdVaJa sc-htpNat sc-htoDjs jplWqw\\" />
                            </StyledComponent>
                          </Styled(Styled(styled.div))>
                        </UsernameInput>
                      </Connect(UsernameInput)>
                    </div>
                  </StyledComponent>
                </Styled(styled.div)>
                <styled.hr borderColor=\\"gray\\" borderWidth={1}>
                  <StyledComponent borderColor=\\"gray\\" borderWidth={1} forwardedComponent={{...}} forwardedRef={{...}}>
                    <hr className=\\"sc-dnqmqq WskOd\\" />
                  </StyledComponent>
                </styled.hr>
                <TodoItem todo={{...}} onChange={[Function]} onDelete={[Function]}>
                  <Styled(styled.div) className=\\"todo-item\\" my=\\"1\\">
                    <StyledComponent className=\\"todo-item\\" my=\\"1\\" forwardedComponent={{...}} forwardedRef={{...}}>
                      <div className=\\"sc-bdVaJa todo-item sc-bwzfXH dKmwgv\\">
                        <Styled(Styled(styled.div)) placeholder=\\"Enter a todo item\\" value=\\"Pending Item\\" onChange={[Function: handleChange]} className=\\"todo-item__input\\" style={{...}} as=\\"input\\" p={2} borderWidth=\\"1px\\" borderStyle=\\"solid\\" borderColor=\\"gray\\" fontSize={1}>
                          <StyledComponent placeholder=\\"Enter a todo item\\" value=\\"Pending Item\\" onChange={[Function: handleChange]} className=\\"todo-item__input\\" style={{...}} as=\\"input\\" p={2} borderWidth=\\"1px\\" borderStyle=\\"solid\\" borderColor=\\"gray\\" fontSize={1} forwardedComponent={{...}} forwardedRef={{...}}>
                            <input placeholder=\\"Enter a todo item\\" value=\\"Pending Item\\" onChange={[Function: handleChange]} className=\\"sc-bdVaJa sc-htpNat todo-item__input sc-htoDjs jplWqw\\" style={{...}} fontSize={1} />
                          </StyledComponent>
                        </Styled(Styled(styled.div))>
                        <Styled(styled.div) onClick={[Function: handleStatusChange]} bg=\\"blue\\" mx=\\"1\\" as=\\"button\\" fontSize=\\"inherit\\" fontWeight=\\"bold\\" m={0} px={3} py={2} color=\\"white\\" border={0} borderRadius={4}>
                          <StyledComponent onClick={[Function: handleStatusChange]} bg=\\"blue\\" mx=\\"1\\" as=\\"button\\" fontSize=\\"inherit\\" fontWeight=\\"bold\\" m={0} px={3} py={2} color=\\"white\\" border={0} borderRadius={4} forwardedComponent={{...}} forwardedRef={{...}}>
                            <button onClick={[Function: handleStatusChange]} fontSize=\\"inherit\\" fontWeight=\\"bold\\" color=\\"white\\" className=\\"sc-bdVaJa sc-EHOje kturus\\">
                              Done
                            </button>
                          </StyledComponent>
                        </Styled(styled.div)>
                        <Styled(styled.div) onClick={[Function: handleDelete]} bg=\\"red\\" as=\\"button\\" fontSize=\\"inherit\\" fontWeight=\\"bold\\" m={0} px={3} py={2} color=\\"white\\" border={0} borderRadius={4}>
                          <StyledComponent onClick={[Function: handleDelete]} bg=\\"red\\" as=\\"button\\" fontSize=\\"inherit\\" fontWeight=\\"bold\\" m={0} px={3} py={2} color=\\"white\\" border={0} borderRadius={4} forwardedComponent={{...}} forwardedRef={{...}}>
                            <button onClick={[Function: handleDelete]} fontSize=\\"inherit\\" fontWeight=\\"bold\\" color=\\"white\\" className=\\"sc-bdVaJa sc-EHOje gPdMQc\\">
                              X
                            </button>
                          </StyledComponent>
                        </Styled(styled.div)>
                      </div>
                    </StyledComponent>
                  </Styled(styled.div)>
                </TodoItem>
                <TodoItem todo={{...}} onChange={[Function]} onDelete={[Function]}>
                  <Styled(styled.div) className=\\"todo-item\\" my=\\"1\\">
                    <StyledComponent className=\\"todo-item\\" my=\\"1\\" forwardedComponent={{...}} forwardedRef={{...}}>
                      <div className=\\"sc-bdVaJa todo-item sc-bwzfXH dKmwgv\\">
                        <Styled(Styled(styled.div)) placeholder=\\"Enter a todo item\\" value=\\"New Item\\" onChange={[Function: handleChange]} className=\\"todo-item__input\\" style={{...}} as=\\"input\\" p={2} borderWidth=\\"1px\\" borderStyle=\\"solid\\" borderColor=\\"gray\\" fontSize={1}>
                          <StyledComponent placeholder=\\"Enter a todo item\\" value=\\"New Item\\" onChange={[Function: handleChange]} className=\\"todo-item__input\\" style={{...}} as=\\"input\\" p={2} borderWidth=\\"1px\\" borderStyle=\\"solid\\" borderColor=\\"gray\\" fontSize={1} forwardedComponent={{...}} forwardedRef={{...}}>
                            <input placeholder=\\"Enter a todo item\\" value=\\"New Item\\" onChange={[Function: handleChange]} className=\\"sc-bdVaJa sc-htpNat todo-item__input sc-htoDjs jplWqw\\" style={{...}} fontSize={1} />
                          </StyledComponent>
                        </Styled(Styled(styled.div))>
                        <Styled(styled.div) onClick={[Function: handleStatusChange]} bg=\\"blue\\" mx=\\"1\\" as=\\"button\\" fontSize=\\"inherit\\" fontWeight=\\"bold\\" m={0} px={3} py={2} color=\\"white\\" border={0} borderRadius={4}>
                          <StyledComponent onClick={[Function: handleStatusChange]} bg=\\"blue\\" mx=\\"1\\" as=\\"button\\" fontSize=\\"inherit\\" fontWeight=\\"bold\\" m={0} px={3} py={2} color=\\"white\\" border={0} borderRadius={4} forwardedComponent={{...}} forwardedRef={{...}}>
                            <button onClick={[Function: handleStatusChange]} fontSize=\\"inherit\\" fontWeight=\\"bold\\" color=\\"white\\" className=\\"sc-bdVaJa sc-EHOje kturus\\">
                              Done
                            </button>
                          </StyledComponent>
                        </Styled(styled.div)>
                        <Styled(styled.div) onClick={[Function: handleDelete]} bg=\\"red\\" as=\\"button\\" fontSize=\\"inherit\\" fontWeight=\\"bold\\" m={0} px={3} py={2} color=\\"white\\" border={0} borderRadius={4}>
                          <StyledComponent onClick={[Function: handleDelete]} bg=\\"red\\" as=\\"button\\" fontSize=\\"inherit\\" fontWeight=\\"bold\\" m={0} px={3} py={2} color=\\"white\\" border={0} borderRadius={4} forwardedComponent={{...}} forwardedRef={{...}}>
                            <button onClick={[Function: handleDelete]} fontSize=\\"inherit\\" fontWeight=\\"bold\\" color=\\"white\\" className=\\"sc-bdVaJa sc-EHOje gPdMQc\\">
                              X
                            </button>
                          </StyledComponent>
                        </Styled(styled.div)>
                      </div>
                    </StyledComponent>
                  </Styled(styled.div)>
                </TodoItem>
                <Styled(Styled(styled.div)) fontSize={3} mt={2} as=\\"h2\\" m={0} fontWeight=\\"bold\\">
                  <StyledComponent fontSize={3} mt={2} as=\\"h2\\" m={0} fontWeight=\\"bold\\" forwardedComponent={{...}} forwardedRef={{...}}>
                    <h2 fontSize={3} fontWeight=\\"bold\\" className=\\"sc-bdVaJa sc-htpNat sc-bxivhb rLXOD\\">
                      Done Items
                    </h2>
                  </StyledComponent>
                </Styled(Styled(styled.div))>
                <TodoItem todo={{...}} onChange={[Function]} onDelete={[Function]}>
                  <Styled(styled.div) className=\\"todo-item\\" my=\\"1\\">
                    <StyledComponent className=\\"todo-item\\" my=\\"1\\" forwardedComponent={{...}} forwardedRef={{...}}>
                      <div className=\\"sc-bdVaJa todo-item sc-bwzfXH dKmwgv\\">
                        <Styled(Styled(styled.div)) placeholder=\\"Enter a todo item\\" value=\\"Done Item\\" onChange={[Function: handleChange]} className=\\"todo-item__input\\" style={{...}} as=\\"input\\" p={2} borderWidth=\\"1px\\" borderStyle=\\"solid\\" borderColor=\\"gray\\" fontSize={1}>
                          <StyledComponent placeholder=\\"Enter a todo item\\" value=\\"Done Item\\" onChange={[Function: handleChange]} className=\\"todo-item__input\\" style={{...}} as=\\"input\\" p={2} borderWidth=\\"1px\\" borderStyle=\\"solid\\" borderColor=\\"gray\\" fontSize={1} forwardedComponent={{...}} forwardedRef={{...}}>
                            <input placeholder=\\"Enter a todo item\\" value=\\"Done Item\\" onChange={[Function: handleChange]} className=\\"sc-bdVaJa sc-htpNat todo-item__input sc-htoDjs jplWqw\\" style={{...}} fontSize={1} />
                          </StyledComponent>
                        </Styled(Styled(styled.div))>
                        <Styled(styled.div) onClick={[Function: handleStatusChange]} bg=\\"salmon\\" mx=\\"1\\" as=\\"button\\" fontSize=\\"inherit\\" fontWeight=\\"bold\\" m={0} px={3} py={2} color=\\"white\\" border={0} borderRadius={4}>
                          <StyledComponent onClick={[Function: handleStatusChange]} bg=\\"salmon\\" mx=\\"1\\" as=\\"button\\" fontSize=\\"inherit\\" fontWeight=\\"bold\\" m={0} px={3} py={2} color=\\"white\\" border={0} borderRadius={4} forwardedComponent={{...}} forwardedRef={{...}}>
                            <button onClick={[Function: handleStatusChange]} fontSize=\\"inherit\\" fontWeight=\\"bold\\" color=\\"white\\" className=\\"sc-bdVaJa sc-EHOje isXsaM\\">
                              Undo
                            </button>
                          </StyledComponent>
                        </Styled(styled.div)>
                        <Styled(styled.div) onClick={[Function: handleDelete]} bg=\\"red\\" as=\\"button\\" fontSize=\\"inherit\\" fontWeight=\\"bold\\" m={0} px={3} py={2} color=\\"white\\" border={0} borderRadius={4}>
                          <StyledComponent onClick={[Function: handleDelete]} bg=\\"red\\" as=\\"button\\" fontSize=\\"inherit\\" fontWeight=\\"bold\\" m={0} px={3} py={2} color=\\"white\\" border={0} borderRadius={4} forwardedComponent={{...}} forwardedRef={{...}}>
                            <button onClick={[Function: handleDelete]} fontSize=\\"inherit\\" fontWeight=\\"bold\\" color=\\"white\\" className=\\"sc-bdVaJa sc-EHOje gPdMQc\\">
                              X
                            </button>
                          </StyledComponent>
                        </Styled(styled.div)>
                      </div>
                    </StyledComponent>
                  </Styled(styled.div)>
                </TodoItem>
              </div>
            </StyledComponent>
          </styled.div>
        </TodoList>
      </div>
    </TodoContainer>
  </Connect(TodoContainer)>
</Provider>"
`;
